### DISCUSSION TITLE
Custom exposure curve for Scuba

### DISCUSSION THREAD

**Discussion by chrisrscuba** [gopro/labs#382] [General]
Hi all,
Sorry in advance for a long post.  Up front, I am not an expert on GoPro Labs, this is really my first effort here.  If you see a way I could be doing this better, pls comment!

I am an avid scuba diver and almost always dive with a GoPro. I like the GoPro because it is simple and small.  But, I have suspected that the exposure curve for the GoPro was biased toward fast shutter speeds and higher ISOs.  For my style of filming when diving, I rarely have fast action but often have very dim light.  This has always resulted in more noise that I would like, especially at depth where there is little light and higher ISOs are required.

I saw the Metadata page and tried MEXPS=1 which shows ISO and shutter speed overlay on the screen.  I looked at these values and discovered that the GoPro does indeed favor higher shutter speeds and therefore higher ISOs than I would like to see in my application.  For example, even with good light I was seeing shutter speeds over 1/300 while the ISO was up at 400.  I would have much preferred lower ISO and slower shutter.  This was even the case with stabilization turned off.  

So, I was looking for a way to modify the exposure curve selection of shutter speed vs ISO.  Note that since the aperture is fixed, this curve basically trades off motion blurriness and stabilization (shutter speed) versus noise.  For underwater applications, the light is often so dim that stabilization is not effective anyway and motion is typically slow.  This is not a knock on the GoPro, it is an action cam so it is understandable that it is biased toward fast shutter speed at the expense of noise, especially since most action applications are outdoors in a well lit environment.  It just does not fit my application for scuba diving.

I read the page on Action Commands / scripting and was intrigued by some of the exposure calculations shown at the bottom using variables and the ISO and shutter speed values.  So, I decided to see how I could modify this curve using scripting and manipulating the shutter speed and Min/Max ISO settings.  I identified 3 areas that I wanted to change based on brightness levels.
   -	Bright – This is for very bright scenes.  In this case I want to lock the Max ISO to 100 and let the shutter speed vary based on scene brightness.  This case ranges from super bright down to an equivalent exposure of when the ISO=100 and the shutter angle is 180 degrees.
   -	Dim – This is for most of my diving (I hope).  In this case, I want to lock the shutter angle to 180 degrees and let the ISO vary based on exposure.  This implies that I will have a “cinematic” shutter angle which just about as slow as it can be (avoiding some of the blurriness) but will also drive the ISO down to a much lower value that the GoPro would have selected on full auto.  This range goes from the ISO=100/180 deg point of the “Bright” range down to ISO=800/180 deg point.  I picked ISO=800 because 1600 starts to be very noisy.  This could be changed depending on taste.
   -	Dark – In this case, it is very dark and my ISO’s are starting to get into the “unusable” range.  In this case, I want to open up the shutter angle to go all the way to 360 degrees and let the ISO vary based on light levels.  In this case, I will let it go all the way to 3200 since if I would prefer to get something, even if it is noisy, than just have black video.  This range is any light level below the “Dim” one.

Calculations
The Action commands page (at the bottom) has a calculation for an EV value.  I don’t need an accurate calculation, just something that is simple to calculate and monotonic.  So, several elements of the equation can be removed to calculate an “Exposure Indicator”.  The equation provided is:
        EV = logbase2 (f-number^2/(time x gain_above_base_iso)) 
I don’t need either the f-number or the logbase2 since they are effectively just scaling the calculation.  

So, I was looking at:
        E=1//(time x gain_above_base_iso))=1/time * 1/ gain_above_base_iso
Explanation of the variables are in the Action command page but the calculation reduced to
=Gi=G*0.01=E1=E/G=E*s.  G and E are variables.  i and s are iso and shutter speed (1/value) numbers.

For my application, I typically shoot with a 30fps shutter speed.  Using this, you can break the above ranges into:
       Bright – E>=60
       Dim – 7.5 <= E  < 60
       Dark – E < 7.5

Conditionals
I tried a variety of conditional structures but had issues in trying to structure it as a series of If, Else If, Else type structures so I had to resort to a series of simple if statements in a sequence that gave me the desired result.  The code became:

If E>=60 then 
shutter angle = auto and ISO Max = 100 (i1M1S0 command)
End If
If E < 60 Then
	  If E >=7.6 Then
	         shutter angle = 180 and ISO Max = 800 (i8M1S180 command)
         End If
End If
If E<7.6 Then
       shutter angle = 360 and ISO Max = 3200 (i32M1S360 command)
End If

I realize that nesting “If” statements would be more efficient but I tried a few combinations and could not get one to work.  Maybe someone here has an approach to restructure this to work more efficiently. 

Note that I had to use 7.6 instead of 7.5 because of the way the conditionals work, I had to have a little buffer to allow the calculation to go below the 7.6  threshold in order to allow the <7.6 conditional to work but the settings for this range prevented the calculation to ever go below 7.5.

So .. Putting it all together, I came up with the following script:
     `=Gi=G*0.01=E1=E/G=E*s>E60i1M1S0<E60>E7.6i8M1S180<E7.6i32M1S360!R`
And it seems to work!!  I have not tried it in the water and still need to do more testing and analysis of recorded clips.  Note, your camera should be in shutter speed =auto and ISO Max/Min = auto.  I have seen some weird behaviors if I had already set limits for those values before running the script.  

Improvements
Some things I may be looking into
     -	May try to reduce the 7.6 number to just over 7.5.  Maybe something like 7.53 This might make a bit smoother transition.
     -	The breakpoints are all based on the frame rate.  So, this script only works for a specific frame rate (30fps).  The frame rate needs to be set prior to running this script.  Either as a preset of using the “pXX” command.  I would like a way to make it frame rate agnostic but this would require being able to read the current frame rate and I have not found out a way to do that.  The alternative is to create separate scripts for all the frame rates I plan to use (likely 24, 30, and 60)
     -	I want to investigate how this would behave if I change it to use 60fps and lock the dim region to 360 degrees instead of 30fps @ 180 degrees.  This would allow 60fps in what I hope is my most used region at a similar noise level but still could support a 2x slow motion in post.  Downside is it will make the “Dark” region more noisy … unless (next item)
     -	Want to see what happens if I attempt to drop frame rate from 60 to 30 (or 24) for dark regions.  Have not tried and have no idea how the camera would react or if will would reject the command.
     -	I am not sure about the repeat rate of the calculations.  I want to experiment with a incrementing loop value to see what the rate is.  I don’t need this to loop super fast, maybe only 2 or 3 times a second (TBD) since it is intended to track gross exposure changes.  I am worried about the impact of too high a calculation rate will have on battery life so need to explore more.  This may just be my lack of knowledge on how these scripts execute and the !R command.
     - Would like to be able to load/execute the script is a specific preset is used and clear it out when other presets are selected.  Can't figure out how to do that.
     - 
Thanks!  Suggestions for improvements welcome!

**Comment by dnewman-gpsw** [comment]
Well I never thought I doing that. I wasn't even sure exposure rules could change mid capture.  I just tried and they do. It is pretty interesting idea.  The camera does some of this already, although I don't know how well it is optimized for dive, so these mode switches may help.

I think this is more reliable:
=Gi=G0.01=E1=E/G=Es>E60i1M1S0+!R>E7.6i8M1S180~i32M1S360!R

Although an issue, once your limit the ISO and shutter, the exposure estimates, will be limited by those settings.  There might not be all the data needed for this. 

P.S. On Mini you should use !MGCMP=1, which disables gyro compensated exposure.  It will came to H11 also.  This helps have a longer shutter in low light.

**Reply by chrisrscuba** [reply]
Thank you for the improved conditional struture.  Never thought of just looping back to the beginning after the first one (did not know you could do that with the !R), will try it.

Any ideas on what the repeat rate is for looping through these types of calculations?

Any ideas on how to load the script based on some camera control/selection?  I would really like to be able to have this script stored and then get the camera to load it (and load a script that clears it out) based on some user action other than the QR codes.  QR Codes are easy and cool but it means you have to carry them around, not a huge deal but one more thing to remember/carry/lose.

Agree on your note about the limits you set on ISO and shutter limiting the exposure calc.  That is why I was using 7.6 instead of 7.5 in the origional structure.  The ISO and shutter limits allow it to go down below 7.6. but need to do more testing to see if there are other issues

**Reply by dnewman-gpsw** [reply]
Processing rate is about 5 milliseconds token; it is intentionally slowed so that Labs processing doesn't hurt normal camera operations.  A token is !thing, or <thing or iThing, the smallest string that is a command.  Some commands need more time to complete, like starting or stopping a capture.  I expect this loop is about 3-4 fps.   Add oMPMSG="loop $l" in the command.  This is useful debug, add you can see how fast it runs.

**Reply by chrisrscuba** [reply]
Thank you, will try it.

**Comment by chrisrscuba** [comment]
Note, I edited the equation in my origional post, for some reason, it was dropping the `<E60>` part in the middle so I had to make it a code snip.  It was there in the edit but not in the preview/final post.  Thinks it is metadata or something and it disappeared even though it was there on the editing page

**Comment by chrisrscuba** [comment]
As an update.  I am using the updated conditionals from dnewman.  Also added a command in the beginning to set ISO/shutter speed to a auto state.  Also addjusted the breakpoint between "Bright" and "Dim" to allow a bit of a smoother transition.  The equation I am using now is:
`i64M1S0=Gi=G*0.01=E1=E/G=E*s>E59i1M1S0+!R7>E7.6i8M1S180~i32M1S360!R7`

Biggest issue is that you can see a Exposure jump when transitioning between regions.  I think it is because of the exposure damping and that when you change the max/min is seems to reset the damping or something.  I want to experiment with the DAMP command and the breakpoint values to experiment

Note:  I am using the GoPro 11 black and the multiplications must have the` '*'` symbol to calculate correctly.  Must use `G*0.01` and `E*s` instead of G0.01 of Es in order for the calculation to work correctl'

**Reply by dnewman-gpsw** [reply]
DAMP will not help, as the camera was not designed to change exposure limits mid-capture, it is only luck that it works at all.

**Reply by chrisrscuba** [reply]
Yeah, probabally but I play with it to see.  I am sure it will always be there, but maybe some tweaking will lessen it.  Since I am only taking periodic clips, there is less chance of it happening during an actual recording and if it does happen during a clip I want to keep, I think I can fix it in post with exposure controls and keyframes when it does happen.  Gotta go diving and try it to see how much of an actual issue it is.  One possible way to avoid would be to lock the exposure during actual recording not sure if that is possible.

**Reply by dnewman-gpsw** [reply]
yes. Add `>r0!R` at the test of your loop.  This way the exposure changes will only happen when not recording.

**Reply by chrisrscuba** [reply]
Oh, that's cool!  Thanks!
This would lock the exposure band/region (Bright, Dim, Dark) during capture but allow the camera to still make exposure adjustments within the settings for that band.  Seems like this would work well!

**Reply by dnewman-gpsw** [reply]
Very happy someone is digging into these new scripting powers.

**Comment by chrisrscuba** [comment]
Having problems savingl/loading the script as a boot script.

Saving:
Used the command `!SAVEscub=i64M1S0=Gi=G*0.01=E1=E/G=E*s>E59i1M1S0+!R7>E7.6i8M1S180~i32M1S360!R7`
What the file actually contained was i64M1S0=Gi=G*0.01=E1=E
The rest was cut off for some reason.  The log file was
[2023-05-26 22:06:11] Read QR:!SAVEscub=i64M1S0=Gi=G*0.01=E1=E/G=E*s>E59i1M1S0+!R7>E7.6i8M1S180~i32M1S360!R7 
[2023-05-26 22:06:11] Save Script:C:\MISC\scub 


Loading
I manually edited the scub file to contain
`i64M1S0=Gi=G*0.01=E1=E/G=E*s>E59i1M1S0+!R7>E7.6i8M1S180~i32M1S360!R7`
and set boot command of `!MBOOT="!Lscub"`

On boot, the script was not working, I don't think it was being read in correctly.  It seems to be partially working but is not acting the same as when I just read in the script and I cannot tell what it is actuall reading in.
The log
[2023-05-26 22:00:47] Labs:Main Start 
[2023-05-26 22:00:47] Labs:Initialize 
[2023-05-26 22:00:48] Boot command:!Lscub 
[2023-05-26 22:00:49] Load Script:C:\MISC\scub 


I am hoping that once I get it working from boot I can get it to work with the QuikCapture

**Reply by dnewman-gpsw** [reply]
This is known bug, fixed in Mini, still pending for H11.  Mini release note: "Fixed using math functions in saved scripts."  The scripts are filtered for potential invalid characters.  In the current release '/', '&', '%' and '|' were incorrectly filtered.  So the =E/G is not working, as the '/' is removed.  So your formula can't use a divide when loaded from an SD card script.  I can't see a workaround at the moment, other then it works as a QR Code.

**Reply by chrisrscuba** [reply]
Thanks, guess I need to wait for a fix.

**Reply by dnewman-gpsw** [reply]
Try this build: https://www.dropbox.com/t/c108Lt1mR0CPRff6   This is not the next release, but should patch the math character error from scripts.

**Reply by chrisrscuba** [reply]
Fantastic!  Works great!  Thank you!  I works on a normal boot but I can't seem to get it to work on QuikCapture.  Are Boot scripts processed for QuikCapture?

**Reply by dnewman-gpsw** [reply]
QC happens first, then any boot scripts.  you have to plan around that in scripting

**Reply by chrisrscuba** [reply]
Not sure what I did wrong but seems to be working now, thanks.  I think maybe the `r>0!R` might have caused it to bypass since it may have already been recording.  Took it out but thought it was still not working but seems to be working.  Need to either live with the bouncing of the brightness or figure another way or delay for a few loops before locking.

**Reply by chrisrscuba** [reply]
That seems to have fixed it.  So, now using `!MBOOT="!Lscub"!SAVEscub=i64M1S0>r0>l3!R7=Gi=G*0.01=E1=E/G=E*s>E59i1M1S0+!R7>E7.6i8M1S180~i32M1S360!R7`

**Reply by chrisrscuba** [reply]
> QC happens first, then any boot scripts. you have to plan around that in scripting

I am still having issues with QC.  It does not seem to run scripts.  To test, I used `'!MBOOT="!Lscub"!SAVEscub="$l"!B0!R` to get a loop count and a beep.  When I do a normal power on, I get a beep and count.  I can then go into record and I still get a beep and count.  If I do a QC, I get nothing.  Not sure why.  Any ideas?

**Reply by chrisrscuba** [reply]
I think I figured it out.  Must have QRDR=1 for some reason.  If have `!MQRDR=1` set, it seems to beep and count down, a bit slow.  If it is not enabled, nothing.

**Reply by dnewman-gpsw** [reply]
Correct

**Comment by chrisrscuba** [comment]
One thing I noticed that may be a bug.  When I disable the script by !MBOOT=0 and restarted the camera, there were still ISO Max limits applied that seemed to be from when the script was running.  When I went into the protune settings they showed Shutter and ISO at auto even though the camera acted as though ISO was set to a Max of 1600.  I set Shutter and ISO to fixed numbers and then back to auto and the limit persisted.  I was able to clear it by scanning a i64M1S0 command.  I would have thought that if there was limits from previous scripting that it would have been reflected in the settings of the preset and you could be able to clear them.  I confirmed this by scanning in a i16M1S0 command and then going into the protune settings.  I showed a limit of 1600.  When I cleard that to auto, the limit at a max of 1600 still persisted even though the protune settings indicated auto.  I tried switching presets, etc. nothing seemed to clear it back except scanning a i64M1S0.  Seems to be for ISO Max only .. ISO Min and Shutter seemed to clear properly.

**Reply by dnewman-gpsw** [reply]
Likely not a bug.  ISO settings are not Labs extensions, your are directly modifying the preset.  Those are sticky. Although I personally don't like auto ISO for Min and Max, as what does that mean?  ISO auto might have a 1600 limit, I don't know.  These are stock camera behaviors.

**Comment by chrisrscuba** [comment]
I made a few changes.  I did not like the way the exposure would bounce back and forth or brighten/darken for a frame or two as you crossed boundaries of the comparisons.  I could have clamped the value during recording but that it problematic for QuikCapture since I think it is already recording when the script starts.  So, I went about trying to figure out a way to directly and set the shutter speed value.  Thru a bunch of back and forth involving EV, shutter speed, and ISO calculations, I came up with an equation with the behavior I like.  Not going to go into details but I can adjust the parameters.  I also reduced the calculation for the "Exposure Indicator" some.  Basically, I am calculating a "Exposure Indicator" similar to the 1st post (just not including the scale factor of `*0.01` and then running that thru an equation of `s=A*Ei+B` type of equation to derive a shutter speed (1/value).  The shutter speed goes to 30 near ISO= 600 and the ISO stays near ISO=200 for most of the EV range.  It goes down to ISO=100 for very bright EV values.

So, I am now using `!MBOOT="!Lscub"!SAVEscub=i64M1S0=Gi=E1=E/G=E*s=E*160=E+24<E31=E30>E870=E870oMEXPX=E=E*1.1oMEXPN=E!R7`

This works, even when recording with QuikCapture.  I do need to test this a lot more and do some capture analysis.

**Comment by chrisrscuba** [comment]
One question.  When I run the above, during boot it goes thru the normal owner info and then indicating what commands are active.  It then  shows "Boot Command: Lscub".  It then clears the "Lscub" part but the "Boot Command:" stays on the screen.  Any way to make it clear out so I can see the whole screen?  If I go into record mode it clears it out and it also clears out when using QuikCapture.
